### ISSUE TITLE
Rewrite front-end to Vue

### ISSUE THREAD

**Issue by christophehenry** [jaywink/socialhome#432]
I'm logging this issue as a follow-up of #202. Now that Streams front-end is almost done, what is the next step?

**Comment by jaywink** [comment]
I think the next step is something like the following:

1) Refactor Vue streams app into a generic frontend app (ie place it under socialhome app, not streams, refactor store, rename many things, etc)
2) Add routing in Vue
3) Add whatever components we need to replace the Django views

I'm not entirely sure how we can do this like we did the streams by using a flag and keeping old functionality working. Maybe if 1) is done first, the streams can still be loaded by creating two bundles, one for just streams functionality, one for new frontend. That would allow slowly building the new frontend and provide a flag to use it, without having to do all things at once quickly. The alternative is a painful to maintain feature branch which needs to have merges done constantly from master. I'm not a huge fan of this idea.

I would keep the frontend in the same repo still. Let's not add too much overhead to the small project :)

Did you have some other thoughts or are we aligned here?

**Comment by christophehenry** [comment]
We're aligned.

**Comment by christophehenry** [comment]
Now giving a second thought, I didn't foresee the whole Vue front would be in its own application. This means the the whole front has to be loaded at once since Webpack and ES6 don't manage conditional imports like Python does. I was expecting we could split the front-end JS code in multiple parts to reduce the amount of uselessly downloaded code. This will ultimately represent a shitload of JS code. We need to find a way to use browser caching at some point.

**Comment by jaywink** [comment]
I don't see the size of the bundle a huge problem. Most of the code is already written. The streams are like 95% of Socialhome. AFAICT it has to be pretty much one bundle so we can make it route nicely in the JS side which makes page transitions faster since there is no page reloading. If we have many bundles and thus independent pages, the transitions are slower, just like currently.

The streams bundle is currently ~500kb or so. That is probably 1/20 of the amount of data used by images in a normal stream refresh. There is nothing we can do to make data usage small unless we start restricting local and remote image loading, which uses many many more times data than our JS does.

**Comment by jaywink** [comment]
I mean if @christophehenry its something you want to do, you could rewrite every single page in vue as separate apps (and bundles), and then we merge them all later into one and add vue routing. Might even be easier than all at once, not sure. Just an option if you want to continue working on the vue side. Though, one could also add stuff to the streams, like filters and content actions.

**Comment by christophehenry** [comment]
Ok. Let's do one bundle and see with time if frontend gets bloated.

**Comment by Quix0r** [comment]
Isn't this a duplicate of #202 ?

**Comment by jaywink** [comment]
Not really, #202 was for rewriting the streams in Vue.js, this is for rewriting the rest of the frontend pages in Vue.js. Most of the frontend is the streams, but there are pages like profile editing, settings, etc, that are still Django templates.

**Comment by Quix0r** [comment]
Why is Django bad for this? I thought vuejs is a client-side template rendering engine. This means that slower computers/laptops/mobiles will take much more time of rendering the actual page than faster ones. If performance is really an issue, then improve Django (e.g. caching of rendered templates so that just the values can be inserted).

**Comment by emacsen** [comment]
@Quix0r Client vs server side rendering is an old discussion at this point. 

When maintaining a large website that is highly responsive with many parts, the client author will end up needing to write a lot of code that handles rendering of status updates, changes to state via websockets, etc.  So it's inevitable that the frontend developer is now forced to control the page anyway, so then it becomes a question of if they will do it ad-hoc or use a framework, like React/Angular or Vue. The answer is that it's generally better to use a framework.

It's also a question of responsiveness, where server-side rendering of an entire page is a very slow process, much slower than you'd get from a single page app. The server may be super fast, but when a web page reloads, the entire HTML page (with all its sub-elements) need to get sent to the browser, which ten needs to render it, including possibly re-fetching all the images or other media. This is what client side rendering is for, so you don't have to re-load the page again. As for caching of rendered templates, that's exactly what client-side rendering allows for. 

Single page apps are what you get on Google Mail, Facebook and other popular websites and services. It doesn't require an especially beefy machine.

**Comment by Quix0r** [comment]
Thank you for explaining it. :+1:
